lib/ClangImporter/ClangImporter.cpp

```diff
@@ -8665,3 +8666,22 @@ importer::getPrivateFileIDAttrs(const clang::Decl *decl) {
 
   return files;
 }
+
+bool importer::declIsCxxOnly(const Decl *decl) {
+  if (auto *clangDecl = decl->getClangDecl()) {
+    return llvm::TypeSwitch<const clang::Decl *, bool>(clangDecl)
+        .template Case<const clang::NamespaceAliasDecl>(
+            [](auto _) { return true; })
+        .template Case<const clang::NamespaceDecl>([](auto _) { return true; })
+        // For the issues this filter function was trying to resolve at its
+        // time of writing, it suffices to only filter out namespaces. But
+        // there are many other kinds of clang::Decls that only appear in C++.
+        // This is obvious for some decls, e.g., templates, using directives,
+        // non-trivial structs, and scoped enums; but it is not obvious for
+        // other kinds of decls, e.g., an enum member or some variable.
+        //
+        // TODO: enumerate those kinds in a precise and robust way
```

Comment:
I don't think we could ever be fully precise just by looking at the AST. People might have `ifdef`s in their code and having arbitrary pieces of code only present in some language modes. Hopefully, most libraries would be well behaved and would not do this. But it would be nice if we had a diagnostic in the future where at least we could tell our users that this was going on (if we can detect this).

Reply:
Yeah, this comment is mostly referring to other clearly C++ decls like template functions and using decls. 

As you point out, it's impossible to reliably warn users either because of ifdefs.